More detailed docs on clock sync

jseldess opened the pull request:
Fixes #960
Fixes #1867
Fixes #2129
Fixes #1799 
Fixes #1690 

@bdarnell, @mberhault, I've added clock sync instructions for digital ocean, gce, aws, azure, and for manual deployment. However, I haven't yet touched on centOS or Redhat, both of which @robert-s-lee has suggested would be helpful. I'm hesitant to document those distros until we've done more testing on them. Please let me know what you think.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/2295)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/7ebd50b7a308602e540e14b20a8de5aa750bbcc5/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/76564397240314573fda0f4b9029726646617073/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/b8c38684a0377a8af3d2d12b6bcaf481604266b9/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/94107c73d9a4fb8192b2ed7872c025c246e1697d/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/57ae057435f58ad2a2c6e1d59e3bb8d676a696a8/

jseldess commented:
Added instructions for GCE as well.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/7f89d92a51c438d4cd8e007d8c24583d3aab5fd0/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/922ee68da8c7702dd4385153926750ea866453bd/

jseldess commented:
Added AWS. Remaining: Azure and manual deployment.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/7d53d9d30c109030e0b9ea53f07ebdac0c3b51db/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9999581568fdb4b999bb1e37e36ccc4b4f49bdf0/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/5e1aeacce110405fb05d31998a040efc70418c43/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/b5a09f443c69e5babf2d7c24610eeda1da188f86/

jseldess commented:
Also added clock offset to the troubleshooting docs.

mberhault commented on _includes/prod_deployment/synchronize-clocks.md:
```diff
@@ -0,0 +1,145 @@
+
+    {% include copy-clipboard.html %}
+    ~~~ shell
+    $ sudo ntpd -b time.google.com
```
`ntpdate`, not `ntpd`. We should also mention why we're doing that.
`ntpdate` also need to be installed, it's not included in the `ntp` package. So the apt-get line becomes: `sudo apt-get install ntpdate ntp`.

If we're using `time.google.com`, then we need to modify `/etc/ntp.conf` as well.

mberhault replied:
Making this change permanent involves changing `/etc/ntp.conf` and restarting the service.
Google [has docs](https://developers.google.com/time/guides#linux_ntpd) on this for various systems.
For ubuntu, the new file looks like:
```
# Specify one or more NTP servers.

# Use servers from the NTP Pool Project. Approved by Ubuntu Technical Board
# on 2011-02-08 (LP: #104525). See http://www.pool.ntp.org/join.html for
# more information.
#pool 0.ubuntu.pool.ntp.org iburst
#pool 1.ubuntu.pool.ntp.org iburst
#pool 2.ubuntu.pool.ntp.org iburst
#pool 3.ubuntu.pool.ntp.org iburst

# Use Ubuntu's ntp server as a fallback.
#pool ntp.ubuntu.com

server time1.google.com iburst
server time2.google.com iburst
server time3.google.com iburst
server time4.google.com iburst
```

Where I removed the `ntp.ubuntu.com` and switched to using the servers specified in the Google docs.

mberhault commented on _includes/prod_deployment/synchronize-clocks.md:
```diff
@@ -0,0 +1,145 @@
+Amazon provides the [Amazon Time Sync Service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/set-time.html), which uses a fleet of satellite-connected and atomic reference clocks in each AWS Region to deliver accurate current time readings. The service also smears the leap second.
+
+- [Configure each AWS instance to use the internal Amazon Time Sync Service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/set-time.html#configure-amazon-time-service).
+- If you plan to run your cluster across cloud providers, [configure the non-AWS machines to use Google's external NTP service](deploy-cockroachdb-on-digital-ocean.html#step-2-sychronize-clocks), which is comparably accurate and also handles <a href="https://developers.google.com/time/smear">"smearing" the leap second</a>.
```
Does amazon not offer an externally-available time server that we can recommended for non-AWS instances?

mberhault commented on _includes/prod_deployment/synchronize-clocks.md:
```diff
@@ -0,0 +1,145 @@
+
+    {% include copy-clipboard.html %}
+    ~~~ shell
+    $ sudo ntpd -b time.google.com
```
ditto: ntpdate.

bdarnell commented:
Review status: 0 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-41:-L0Rswr6IgUAW_0zhmMU:b-oreaky) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L41)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

`ntpdate`, not `ntpd`. We should also mention why we're doing that.
`ntpdate` also need to be installed, it's not included in the `ntp` package. So the apt-get line becomes: `sudo apt-get install ntpdate ntp`.

If we're using `time.google.com`, then we need to modify `/etc/ntp.conf` as well.
</blockquote></details>

Any discussion of time.google.com needs to be accompanied by a warning that if any of your nodes use google's time server, they all must. Reconfiguring servers to use google's time service is more involved than simply enabling ntp with the default pool, so I'm not sure if we want to recommend this by default (OTOH, unless you do something about smearing the cluster will crash on the leap second)

---

*[_includes/prod_deployment/synchronize-clocks.md, line 74 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-74:-L0RuFWnMFTH2_6NKrto:b-7yvhlx) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L74)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Does amazon not offer an externally-available time server that we can recommended for non-AWS instances?
</blockquote></details>

I don't think they do. 

---

*[_includes/prod_deployment/synchronize-clocks.md, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0RtjjfzbcD8YIoTSET:-L0RtjjfzbcD8YIoTSEU:b-sjp2tl) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L86)):*
> ```Markdown
>     {% include copy-clipboard.html %}
>     ~~~ shell
>     $ curl -O https://raw.githubusercontent.com/torvalds/linux/master/tools/hv/lsvmbus
> ```

Suggesting that users copy these commands one at a time is awkward. We should be providing a script that can be run in a single shot. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[_includes/prod_deployment/synchronize-clocks.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-41:-L0UjW_7tRHFHg1Tvorg:b-mez1nk) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L41)):*
Thanks, @mberhault and @bdarnell. I'll make changes soon. But @mberhault, why is it important to use `ntpdate` as opposed to `ntpd`? The [docs](http://doc.ntp.org/4.1.1/ntpdate.htm) say:

> Disclaimer: The functionality of this program is now available in the ntpd program. See the -q command line option in the ntpd - Network Time Protocol (NTP) daemon page. After a suitable period of mourning, the ntpdate program is to be retired from this distribution

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[_includes/prod_deployment/synchronize-clocks.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-41:-L0UkPkFultKtPjysvKj:b-k8xjsl) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L41)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Thanks, @mberhault and @bdarnell. I'll make changes soon. But @mberhault, why is it important to use `ntpdate` as opposed to `ntpd`? The [docs](http://doc.ntp.org/4.1.1/ntpdate.htm) say:

> Disclaimer: The functionality of this program is now available in the ntpd program. See the -q command line option in the ntpd - Network Time Protocol (NTP) daemon page. After a suitable period of mourning, the ntpdate program is to be retired from this distribution
</blockquote></details>

Also, @mberhault, if we say to modify `/etc/ntp.conf`, when would they do that? After or before the steps here? And would we just tell them to do [this](https://developers.google.com/time/guides#linux_ntpd)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[_includes/prod_deployment/synchronize-clocks.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-41:-L0UkvvJUAoyWZu3k6aQ:b9xse2b) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L41)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Also, @mberhault, if we say to modify `/etc/ntp.conf`, when would they do that? After or before the steps here? And would we just tell them to do [this](https://developers.google.com/time/guides#linux_ntpd)?
</blockquote></details>

@bdarnell, are you ok with our recommending google's time server? Seems easiest way to account for leap second smearing. If so, you mention that the process is more involved. What's missing here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 0 of 19 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-41:-L0Ul0bsw_3dc6-LXD6-:b-bpf9ay) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L41)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

@bdarnell, are you ok with our recommending google's time server? Seems easiest way to account for leap second smearing. If so, you mention that the process is more involved. What's missing here?
</blockquote></details>

oh, yeah, it's `ntpd -q` (just set the time and quit). Definitely no `-b` though, that's a `ntpdate` flag.
If we're using custom servers to `ntpd -q`, then we need to make sure those same servers are in the config file. It needs to be done after `ntpd` is stopped and before it is started up again.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 19 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 9 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0WGAI7mUP6oj25VUod:-L0WGAI7mUP6oj25VUoe:b-nij5i0) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L9)):*
> ```Markdown
> 1. SSH to the first machine.
> 
> 2. Disable `timesyncd`, which tends to be active by default on some Linux distributions:
> ```

`timesyncd` is the default on ubuntu. Could/should we recommend that ubuntu users point `timesyncd` at `time.google.com` instead of switching to `ntpd`? Is the preference for `ntpd` just because that's the one we're most familiar with or is there some reason to favor it over other NTP clients?

---

*[_includes/prod_deployment/synchronize-clocks.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-41:-L0WDvcOZqK7sgevTjsz:b773l3o) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L41)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

oh, yeah, it's `ntpd -q` (just set the time and quit). Definitely no `-b` though, that's a `ntpdate` flag.
If we're using custom servers to `ntpd -q`, then we need to make sure those same servers are in the config file. It needs to be done after `ntpd` is stopped and before it is started up again.
</blockquote></details>

The "more involved" part is the fact that you need to edit `/etc/ntp.conf` and *make sure* that you do this for every current and future node. I agree that this is probably still the simplest way to account for the leap second, though. (since to do client-side slewing it looks like you need to use chronyd instead of ntpd)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[_includes/prod_deployment/synchronize-clocks.md, line 9 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0WGAI7mUP6oj25VUod:-L0XN3S1WyWMCWyUKqYx:b-swq3wq) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L9)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`timesyncd` is the default on ubuntu. Could/should we recommend that ubuntu users point `timesyncd` at `time.google.com` instead of switching to `ntpd`? Is the preference for `ntpd` just because that's the one we're most familiar with or is there some reason to favor it over other NTP clients?
</blockquote></details>

Good question. I've read here and here that `timesyncd` should be sufficient. I could have the user check for `timesyncd` and, if it's present, configure it, if it's not present, install and configure `ntp`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[_includes/prod_deployment/synchronize-clocks.md, line 9 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0WGAI7mUP6oj25VUod:-L0XSjOuzYDZGVItSPl2:b-yd0znh) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L9)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Good question. I've read here and here that `timesyncd` should be sufficient. I could have the user check for `timesyncd` and, if it's present, configure it, if it's not present, install and configure `ntp`?
</blockquote></details>

I've also read that [`chrony` performs better](https://chrony.tuxfamily.org/comparison.html). And [configuration seems just as easy]( https://www.server-world.info/en/note?os=Ubuntu_16.04&p=ntp&f=2).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@mberhault, @bdarnell, any updates/thoughts on this work? I believe Mark was going to try to use ntp with the external google time service on one of our production clusters before we publish these recommendations...

mberhault commented on _includes/prod_deployment/synchronize-clocks.md:
```diff
@@ -0,0 +1,145 @@
+
+    {% include copy-clipboard.html %}
+    ~~~ shell
+    $ sudo timedatectl set-ntp no
```
I can confirm that this disables `systemd-timesyncd.service` permanently (as opposed to just stopping it). This is mentioned in the [timedatectl docs](https://www.freedesktop.org/software/systemd/man/timedatectl.html):
```
This command is hence mostly equivalent to: systemctl enable --now systemd-timesyncd.service
and systemctl disable --now systemd-timesyncd.service, but is protected by a different access policy.
```

jseldess replied:
Great, thanks.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/084348c3bd93c204e72725a9ed75b38d603195ef/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/0abd007a6ed17bbe8df527472475f6c183fb2786/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/2487ba91077f0cc496a7c20eab28f9f06f88830b/

jseldess commented:
Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 41 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-41:-L2ewvNnj7xt3VsWWe7S:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L41)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Making this change permanent involves changing `/etc/ntp.conf` and restarting the service.
Google [has docs](https://developers.google.com/time/guides#linux_ntpd) on this for various systems.
For ubuntu, the new file looks like:
```
# Specify one or more NTP servers.

# Use servers from the NTP Pool Project. Approved by Ubuntu Technical Board
# on 2011-02-08 (LP: #104525). See http://www.pool.ntp.org/join.html for
# more information.
#pool 0.ubuntu.pool.ntp.org iburst
#pool 1.ubuntu.pool.ntp.org iburst
#pool 2.ubuntu.pool.ntp.org iburst
#pool 3.ubuntu.pool.ntp.org iburst

# Use Ubuntu's ntp server as a fallback.
#pool ntp.ubuntu.com

server time1.google.com iburst
server time2.google.com iburst
server time3.google.com iburst
server time4.google.com iburst
```

Where I removed the `ntp.ubuntu.com` and switched to using the servers specified in the Google docs.
</blockquote></details>

Done.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0RtjjfzbcD8YIoTSET:-L0UjzX4PCt8V9XcKM5O:bm7eq7a) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L86)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Suggesting that users copy these commands one at a time is awkward. We should be providing a script that can be run in a single shot. 
</blockquote></details>

Marc, Ben, I think the instructions are now up to date with our latest understanding. Do you think it's still meaningful to write script? They have to edit the `conf` file, which might need to be done, manually, right?

---

*[_includes/prod_deployment/synchronize-clocks.md, line 124 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0NkYiXba0uM-fx3Teu-r3-124:-L2exLSYHSmS-yclFBt0:b-oif3on) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L124)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

ditto: ntpdate.
</blockquote></details>

I can't find where, but I'm pretty sure we agreed to go with `ntpd` instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@mberhault, @bdarnell, the docs should be up-to-date with our latest discussions. Please take another look before I apply the changes to the 2.0 docs as well. Thanks for all the help here!

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/30261bf9fa486c14d3fd8397fec9f53921af7933/

bdarnell commented:
Reviewed 2 of 6 files at r4.
Review status: 2 of 13 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[_includes/faq/clock-synchronization.html, line 1 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2fefEIX1nKMwmnlo4u:-L2fefEIX1nKMwmnlo4v:b-fgdmrk) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/faq/clock-synchronization.html#L1)):*
> ```HTML
> CockroachDB requires moderate levels of clock synchronization to preserve data consistency. For this reason, when a node detects that its clock is out of synch with at least half of the other nodes in the cluster by 80% of the maximum offset allowed (500ms by default), it spontaneously shuts down. This avoids the risk of violating [serializable consistency](https://en.wikipedia.org/wiki/Serializability) and causing stale reads and write skews, but it's best to prevent clocks from drifting too far in the first place by running [NTP](http://www.ntp.org/) or other clock synchronization software on each node.
> ```

I think "it's best" is too weak here - I'd say "it's important"

---

*[_includes/prod_deployment/synchronize-clocks.md, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0RtjjfzbcD8YIoTSET:-L2ffwXC0I8MiITZBDyV:bfcz6r7) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L86)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Marc, Ben, I think the instructions are now up to date with our latest understanding. Do you think it's still meaningful to write script? They have to edit the `conf` file, which might need to be done, manually, right?
</blockquote></details>

Yes, I still think we should write a script for this.

The conf file would not be edited manually in practice - you'd have a master copy in some config repository and copy it over. 

---

*[_includes/prod_deployment/synchronize-clocks.md, line 5 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2fevbFByTFSiB8kyrO:-L2fevbFByTFSiB8kyrP:b90qovm) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L5)):*
> ```Markdown
> {% if page.title contains "Digital Ocean" or page.title contains "Manual" %}
> 
> [NTP](http://www.ntp.org/) should keep offsets in the single-digit milliseconds, so that software is featured here, but other methods of clock synchronization are suitable as well.
> ```

The software we're featuring here is `ntpd`, not NTP. NTP is the protocol, which is used by all three software packages (ntpd, timesyncd, chrony). 

---

*[_includes/prod_deployment/synchronize-clocks.md, line 46 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2ffIA5MmpkklQzhDGA:-L2ffIA5MmpkklQzhDGB:b-3e0qk5) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L46)):*
> ```Markdown
>     ~~~
> 
>     To make this change permanent, in the `/etc/ntp.conf` file, remove or comment out any lines starting with `server` and add the following lines:
> ```

From Marc's previous comment, it looks like you also have to remove any `pool` lines as well. 

Is there anything in the file that we would want to keep, or can we just say to replace the entire file with these four `server` lines?

---

*[_includes/prod_deployment/synchronize-clocks.md, line 78 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2ffU8DP9DUWQ4Egzji:-L2ffU8DP9DUWQ4Egzjj:b-pmgds) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L78)):*
> ```Markdown
> {% elsif page.title contains "Google" %}
> 
> Compute Engine instances are preconfigured to use [NTP](http://www.ntp.org/), which should keep offsets in the single-digit milliseconds. However, Google can’t predict how external NTP services, such as `pool.ntp.org`, will handle the leap second. Therefore, you should:
> ```

Whether they are pre-configured with NTP depends on which base image you're using; I don't think this is guaranteed. 

---

*[_includes/prod_deployment/synchronize-clocks.md, line 88 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2ffhQVzs2CvnBT-1Ex:-L2ffhQVzs2CvnBT-1Ey:bmqhwuc) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L88)):*
> ```Markdown
> 
> - [Configure each AWS instance to use the internal Amazon Time Sync Service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/set-time.html#configure-amazon-time-service).
> - If you plan to run your cluster across cloud providers, [configure the non-AWS machines to use Google's external NTP service](deploy-cockroachdb-on-digital-ocean.html#step-2-sychronize-clocks), which is comparably accurate and also handles <a href="https://developers.google.com/time/smear">"smearing" the leap second</a>.
> ```

I don't think it's safe to mix amazon's internal time servers with google's external ones. I think if you envision a cross-cloud deployment you should use google's external time servers even on AWS. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Note: Mention somewhere that you can get the current cluster offset via the `clock_offset_* variables` at `/_status/vars`.

mberhault commented:
Review status: 2 of 13 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 46 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2ffIA5MmpkklQzhDGA:-L34Sklv2yZrsVbh4RYG:b4ayjl5) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L46)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

From Marc's previous comment, it looks like you also have to remove any `pool` lines as well. 

Is there anything in the file that we would want to keep, or can we just say to replace the entire file with these four `server` lines?
</blockquote></details>

Yup, the pool lines all need to go away. I kept the rest of the file as the GCE machines had them, but I'm not sure which parts are useful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 2 of 13 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[_includes/faq/clock-synchronization.html, line 1 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2fefEIX1nKMwmnlo4u:-L34LwqyO_eHUoeZXPZM:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/faq/clock-synchronization.html#L1)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think "it's best" is too weak here - I'd say "it's important"
</blockquote></details>

Done.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0RtjjfzbcD8YIoTSET:-L34Tk6EEhS8VG1me31F:b-mlqn2r) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L86)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yes, I still think we should write a script for this.

The conf file would not be edited manually in practice - you'd have a master copy in some config repository and copy it over. 
</blockquote></details>

I'd like to get this merged, so I'd prefer to convert these steps to a script in a follow-up PR. 

---

*[_includes/prod_deployment/synchronize-clocks.md, line 5 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2fevbFByTFSiB8kyrO:-L34Pz9epqqRESGKGClA:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L5)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

The software we're featuring here is `ntpd`, not NTP. NTP is the protocol, which is used by all three software packages (ntpd, timesyncd, chrony). 
</blockquote></details>

Done.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 78 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2ffU8DP9DUWQ4Egzji:-L34Qagp8Qw4eoUHAHmP:b-5cbul4) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L78)):*
The [GCE docs](https://cloud.google.com/compute/docs/instances/managing-instances#configure-ntp) claim:

> To help keep system clocks in sync, Compute Engine instances are preconfigured to use network time protocol (NTP).

So I'd prefer to trust them and keep this simple until someone tells us it's wrong. That ok?

---

*[_includes/prod_deployment/synchronize-clocks.md, line 88 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2ffhQVzs2CvnBT-1Ex:-L34S84PpISmpS51pjJh:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L88)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I don't think it's safe to mix amazon's internal time servers with google's external ones. I think if you envision a cross-cloud deployment you should use google's external time servers even on AWS. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[_includes/prod_deployment/synchronize-clocks.md, line 46 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2ffIA5MmpkklQzhDGA:-L34UOvYdpXjwhHzeoKq:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L46)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

Yup, the pool lines all need to go away. I kept the rest of the file as the GCE machines had them, but I'm not sure which parts are useful.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f445384a38cabcc68db72fe1dd8459f05428058a/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/879fd189ee743de2520e058fd27518d4f46a9cf6/

mberhault commented:
Review status: 0 of 24 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0RtjjfzbcD8YIoTSET:-L34Z8AUAI9qRrN5Ni0p:b-2r03bl) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L86)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

I'd like to get this merged, so I'd prefer to convert these steps to a script in a follow-up PR. 
</blockquote></details>

I think a single copyable text block would do rather than individual blocks (for single-shot copy/paste). Description of what it does can be moved above. That should be just for the Azure bits, the ntp switcheroo can remain individual blocks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 14 files at r5.
Review status: 1 of 24 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[_includes/prod_deployment/synchronize-clocks.md, line 78 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L2ffU8DP9DUWQ4Egzji:-L34bfTZDndm0bJt94_q:b32w0x) ([raw file](https://github.com/cockroachdb/docs/blob/30261bf9fa486c14d3fd8397fec9f53921af7933/_includes/prod_deployment/synchronize-clocks.md#L78)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

The [GCE docs](https://cloud.google.com/compute/docs/instances/managing-instances#configure-ntp) claim:

> To help keep system clocks in sync, Compute Engine instances are preconfigured to use network time protocol (NTP).

So I'd prefer to trust them and keep this simple until someone tells us it's wrong. That ok?
</blockquote></details>

OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->

mberhault commented:
LGTM. Let's leave scripting for later if we need it. We would have to account for various systems before providing a full copy/paste solution.

---

Review status: 1 of 24 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-:-L34eFnQDzQ-KZjE0fPb:b-l7vfmd)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/8247cda8067244588c01aa9a9c3397a44d36bfe3/

jseldess commented:
*[_includes/prod_deployment/synchronize-clocks.md, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/2295#-L0RtjjfzbcD8YIoTSET:-L34if9frRTIfQmZrcPe:bq51y3y) ([raw file](https://github.com/cockroachdb/docs/blob/b5a09f443c69e5babf2d7c24610eeda1da188f86/_includes/prod_deployment/synchronize-clocks.md#L86)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I think a single copyable text block would do rather than individual blocks (for single-shot copy/paste). Description of what it does can be moved above. That should be just for the Azure bits, the ntp switcheroo can remain individual blocks.
</blockquote></details>

We discussed offline. Given that unbinding the device relies on retrieving the DEVICE_ID from the output of the second command, and it's not guaranteed to be identical on all system, we agreed to keep the instructions explicit for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/2295)*
<!-- Sent from Reviewable.io -->
